TombEngine/Objects/TR3/Entity/tr3_civvy.cpp

```diff
@@ -2,418 +2,594 @@
-					extraTorsoRot.x = AI.xAngle;
-					extraTorsoRot.y = AI.angle;
-				}
+					break;
 
-				if (AI.bite && AI.distance < CIVVY_ATTACK2_RANGE)
-					item->Animation.TargetState = CIVVY_PUNCH2;
-				else
-					item->Animation.TargetState = CIVVY_STATE_WALK_FORWARD;
+				case CIVVY_STATE_AIM_FAR_PUNCH:
 
-				break;
+					creature.MaxTurn = CIVVY_AIM_TURN_RATE_MAX;
+					creature.Flags = 0;
 
-			case CIVVY_PUNCH0:
-				creature->MaxTurn = CIVVY_WALK_TURN_RATE_MAX;
+					if (AI.ahead)
+					{
+						jointTorsoRot.x = AI.xAngle;
+						jointTorsoRot.y = AI.angle;
+					}
 
-				if (AI.ahead)
-				{
-					extraTorsoRot.x = AI.xAngle;
-					extraTorsoRot.y = AI.angle;
-				}
+					if (creature.Enemy != nullptr && AI.bite && AI.distance < CIVVY_ATTACK_FAR_PUNCH_RANGE)
+						item.Animation.TargetState = CIVVY_STATE_ATTACK_FAR_PUNCH;
+					else
+						item.Animation.TargetState = CIVVY_STATE_IDLE;
 
-				if (!creature->Flags && item->TouchBits.Test(CivvyAttackJoints))
-				{
-					CreatureEffect(item, CivvyBite, DoBloodSplat);
-					DoDamage(creature->Enemy, CIVVY_ATTACK_DAMAGE);
-					SoundEffect(SFX_TR4_LARA_THUD, &item->Pose);
-					creature->Flags = 1;
-				}
+					break;
 
-				break;
+				case CIVVY_STATE_AIM_WALKING_PUNCH:
 
-			case CIVVY_PUNCH1:
-				creature->MaxTurn = CIVVY_WALK_TURN_RATE_MAX;
+					creature.MaxTurn = CIVVY_AIM_TURN_RATE_MAX;
+					creature.Flags = 0;
 
-				if (AI.ahead)
-				{
-					extraTorsoRot.x = AI.xAngle;
-					extraTorsoRot.y = AI.angle;
-				}
+					if (AI.ahead)
+					{
+						jointTorsoRot.x = AI.xAngle;
+						jointTorsoRot.y = AI.angle;
+					}
 
-				if (!creature->Flags && item->TouchBits.Test(CivvyAttackJoints))
-				{
-					CreatureEffect(item, CivvyBite, DoBloodSplat);
-					DoDamage(creature->Enemy, CIVVY_ATTACK_DAMAGE);
-					SoundEffect(SFX_TR4_LARA_THUD, &item->Pose);
-					creature->Flags = 1;
-				}
+					
+					if (creature.Enemy != nullptr && AI.bite && AI.distance < CIVVY_ATTACK_WALKING_PUNCH_RANGE)
+						item.Animation.TargetState = CIVVY_STATE_ATTACK_WALKING_PUNCH;
+					else
+						item.Animation.TargetState = CIVVY_STATE_WALK;
+
+					break;
 
-				if (AI.ahead && AI.distance > CIVVY_ATTACK1_RANGE && AI.distance < CIVVY_ATTACK2_RANGE)
-					item->Animation.TargetState = CIVVY_PUNCH2;
+				case CIVVY_STATE_ATTACK_CLOSE_PUNCH:
 
-				break;
+					creature.MaxTurn = CIVVY_WALK_TURN_RATE_MAX;
 
-			case CIVVY_PUNCH2:
-				creature->MaxTurn = CIVVY_WALK_TURN_RATE_MAX;
+					if (AI.ahead)
+					{
+						jointTorsoRot.x = AI.xAngle;
+						jointTorsoRot.y = AI.angle;
+					}
 
-				if (AI.ahead)
-				{
-					extraTorsoRot.x = AI.xAngle;
-					extraTorsoRot.y = AI.angle;
-				}
+					if (creature.Flags == 0)
+					{
+						if (creature.Enemy->IsLara())
+						{
+							if (item.TouchBits.Test(CivvyAttackJoints))
+								creature.Flags = 2;
+						}
+						else
+						{
+							float distance = Vector3i::Distance(item.Pose.Position, creature.Enemy->Pose.Position);
+							if (distance <= CLICK(2))
```

Comment:
Since distance is not a height, it can simply use `BLOCK(0.5f)` here. :)

Reply:
Done